R/linear_pool.R

```diff
@@ -0,0 +1,173 @@
+#' @param ... parameters that are passed to `distfromq::make_r_fun`, specifying
+#'   details of how to estimate a quantile function from provided quantile levels and
+#'.  quantile values.
+#'
+#' @return a `model_out_tbl` object of ensemble predictions. Note that
+#'   any additional columns in the input `model_outputs` are dropped.
+#'
+#' @export
+linear_pool <- function(model_outputs, weights = NULL,
+                        weights_col_name = "weight",
+                        model_id = "hub-ensemble",
+                        task_id_cols = NULL,
+                        ...) {
+                                
+  if (!is.data.frame(model_outputs)) {
+    cli::cli_abort(c("x" = "{.arg model_outputs} must be a `data.frame`."))
+  }
+
+  if (isFALSE("model_out_tbl" %in% class(model_outputs))) {
+    model_outputs <- hubUtils::as_model_out_tbl(model_outputs)
+  }
+
+  model_out_cols <- colnames(model_outputs)
+
+  non_task_cols <- c("model_id", "output_type", "output_type_id", "value")
+  if (is.null(task_id_cols)) {
+    task_id_cols <- model_out_cols[!model_out_cols %in% non_task_cols]
+  }
+
+  if (!all(task_id_cols %in% model_out_cols)) {
+    cli::cli_abort(c(
+      "x" = "{.arg model_outputs} did not have all listed task id columns
+             {.val {task_id_col}}."
+    ))
+  }
+  
+  # check `model_outputs` has all standard columns with correct data type
+  # and `model_outputs` has > 0 rows
+  hubUtils::validate_model_out_tbl(model_outputs)
+
+  valid_types <- c("mean", "quantile", "cdf", "pmf", "sample")
+  unique_types <- unique(model_outputs[["output_type"]])
+  invalid_types <- unique_types[!unique_types %in% valid_types]
+  if (length(invalid_types) > 0) {
+    cli::cli_abort(c(
+      "x" = "{.arg model_outputs} contains unsupported output type.",
+      "!" = "Included invalid output type{?s}: {.val {invalid_types}}.",
+      "i" = "Supported output types: {.val {valid_types}}."
+    ))
+  }
+
+  # calculate linear opinion pool for different types
+  ensemble_outputs1 <- ensemble_outputs2 <- ensemble_outputs3 <- NULL
+  
+  if (any(unique_types %in% c("mean", "cdf", "pmf"))) {
+    # linear pool calculation for mean, cdf, pmf output types
+    ensemble_outputs1 <- model_outputs %>%
+      dplyr::filter(output_type %in% c("mean", "cdf", "pmf")) %>%
+      hubEnsembles::simple_ensemble(weights = weights,
+                                    weights_col_name = weights_col_names,
+                                    agg_fun = "mean", agg_args = list(),
+                                    model_id = model_id,
+                                    task_id_cols = task_id_cols) 
+  } 
+  
+  if (any(unique_types == "sample")) {
+    # linear pool calculation for sample output type
+    print("sample")
+  } 
+  
+  if (any(unique_types == "quantile")) {
+    # linear pool calculation for quantile output type
+    n_samples <- 1e4
+    quantile_levels <- unique(model_outputs$output_type_id)
+    
+    if (is.null(weights)) {
```

Comment:
Since most of the code in lines 98 through 145 got moved to the new `validate_ensemble_inputs` and there are now only a few duplicated lines between the two ensemble functions, I don't think it's necessary to make another function. Let me know if you think otherwise, though